service/src/main/java/bio/terra/axonserver/utils/CloudStorageUtils.java

```diff
@@ -70,15 +77,32 @@ public static byte[] getBucketObject(
             .getService()
             .reader(blobId)) {
 
+      long startByteIdx =
+          Optional.ofNullable(byteRange).isPresent() ? byteRange.getRangeStart(Long.MAX_VALUE) : 0;
+      long endByteIdx =
+          Optional.ofNullable(byteRange).isPresent()
+              ? byteRange.getRangeEnd(Long.MAX_VALUE)
+              : Long.MAX_VALUE - 1;
+      long bytesToRead = endByteIdx - startByteIdx;
+
       BoundedByteArrayOutputStream outputStream = new BoundedByteArrayOutputStream(MAX_OBJECT_SIZE);
-      ByteBuffer bytes = ByteBuffer.allocate(MAX_BUFFER_SIZE);
-      while (reader.read(bytes) > 0) {
-        bytes.flip();
-        outputStream.write(bytes.array(), 0, bytes.limit());
-        bytes.clear();
+      ByteBuffer buffer = ByteBuffer.allocate(MAX_BUFFER_SIZE);
+
+      long totalBytesRead = 0;
+      int bytesRead;
+      reader.seek(startByteIdx);
+      while (totalBytesRead < bytesToRead && reader.read(buffer) > 0) {
+        bytesRead = buffer.position();
```

Comment:
did you test this?

i don't quiet follow the logic of the while loop here.
in line 100, buffer.clear() will reset the buffer's position to 0.
byteread will be 0 at line 95, totalBytesRead will not increment i think?

Reply:
reader.read(buffer) in the while loop condition on like 94 will fill the buffer at the start of each iteration. bytesRead will either be the size of the buffer or the number of remaining bytes if the reader channel reaches the end of stream.